File: src-wails/dirty_manager.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"sync"
	"time"
)

type DirtyItem struct {
	ID           string    `json:"id"`
	Type         string    `json:"type"`
	OriginalHash string    `json:"originalHash"`
	CurrentHash  string    `json:"currentHash"`
	IsDirty      bool      `json:"isDirty"`
	LastModified time.Time `json:"lastModified"`
	LastSaved    time.Time `json:"lastSaved"`
}

type DirtyManager struct {
	items map[string]*DirtyItem
	mu    sync.RWMutex
}

func NewDirtyManager() *DirtyManager {
	return &DirtyManager{
		items: make(map[string]*DirtyItem),
	}
}

func computeHash(content []byte) string {
	hash := sha256.Sum256(content)
	return hex.EncodeToString(hash[:])
}

func (dm *DirtyManager) SetOriginal(id, itemType string, content []byte) {
	dm.mu.Lock()
	defer dm.mu.Unlock()
	hash := computeHash(content)
	dm.items[id] = &DirtyItem{
		ID:           id,
		Type:         itemType,
		OriginalHash: hash,
		CurrentHash:  hash,
		IsDirty:      false,
		LastModified: time.Now(),
		LastSaved:    time.Now(),
	}
}

func (dm *DirtyManager) UpdateCurrent(id string, content []byte) bool {
	dm.mu.Lock()
	defer dm.mu.Unlock()
	item, exists := dm.items[id]
	if !exists {
		return false
	}
	newHash := computeHash(content)
	if newHash != item.CurrentHash {
		item.CurrentHash = newHash
		item.IsDirty = newHash != item.OriginalHash
		item.LastModified = time.Now()
		return true
	}
	return false
}

func (dm *DirtyManager) MarkSaved(id string, content []byte) {
	dm.mu.Lock()
	defer dm.mu.Unlock()
	item, exists := dm.items[id]
	if !exists {
		return
	}
	hash := computeHash(content)
	item.OriginalHash = hash
	item.CurrentHash = hash
	item.IsDirty = false
	item.LastSaved = time.Now()
}

func (dm *DirtyManager) IsDirty(id string) bool {
	dm.mu.RLock()
	defer dm.mu.RUnlock()
	item, exists := dm.items[id]
	if !exists {
		return false
	}
	return item.IsDirty
}

func (dm *DirtyManager) GetItem(id string) *DirtyItem {
	dm.mu.RLock()
	defer dm.mu.RUnlock()
	item, exists := dm.items[id]
	if !exists {
		return nil
	}
	itemCopy := *item
	return &itemCopy
}

func (dm *DirtyManager) GetDirtyItems() []DirtyItem {
	dm.mu.RLock()
	defer dm.mu.RUnlock()
	var result []DirtyItem
	for _, item := range dm.items {
		if item.IsDirty {
			result = append(result, *item)
		}
	}
	return result
}

func (dm *DirtyManager) GetAllItems() []DirtyItem {
	dm.mu.RLock()
	defer dm.mu.RUnlock()
	var result []DirtyItem
	for _, item := range dm.items {
		result = append(result, *item)
	}
	return result
}

func (dm *DirtyManager) ClearItem(id string) {
	dm.mu.Lock()
	defer dm.mu.Unlock()
	delete(dm.items, id)
}

func (dm *DirtyManager) HasAnyDirty() bool {
	dm.mu.RLock()
	defer dm.mu.RUnlock()
	for _, item := range dm.items {
		if item.IsDirty {
			return true
		}
	}
	return false
}

func (dm *DirtyManager) GetMinutesSinceLastSave(id string) int {
	dm.mu.RLock()
	defer dm.mu.RUnlock()
	item, exists := dm.items[id]
	if !exists {
		return 0
	}
	return int(time.Since(item.LastSaved).Minutes())
}

type DirtyState struct {
	ID                  string `json:"id"`
	IsDirty             bool   `json:"isDirty"`
	MinutesSinceLastSave int   `json:"minutesSinceLastSave"`
}

func (a *App) GetDirtyState(id string) *DirtyState {
	if a.dirtyManager == nil {
		return nil
	}
	item := a.dirtyManager.GetItem(id)
	if item == nil {
		return nil
	}
	return &DirtyState{
		ID:                  id,
		IsDirty:             item.IsDirty,
		MinutesSinceLastSave: a.dirtyManager.GetMinutesSinceLastSave(id),
	}
}

func (a *App) SetOriginalContent(id, itemType, content string) {
	if a.dirtyManager == nil {
		a.dirtyManager = NewDirtyManager()
	}
	a.dirtyManager.SetOriginal(id, itemType, []byte(content))
}

func (a *App) UpdateCurrentContent(id, content string) bool {
	if a.dirtyManager == nil {
		return false
	}
	return a.dirtyManager.UpdateCurrent(id, []byte(content))
}

func (a *App) MarkContentSaved(id, content string) {
	if a.dirtyManager == nil {
		return
	}
	a.dirtyManager.MarkSaved(id, []byte(content))
}

func (a *App) IsContentDirty(id string) bool {
	if a.dirtyManager == nil {
		return false
	}
	return a.dirtyManager.IsDirty(id)
}

func (a *App) HasAnyDirtyContent() bool {
	if a.dirtyManager == nil {
		return false
	}
	return a.dirtyManager.HasAnyDirty()
}

func (a *App) GetAllDirtyStates() []DirtyState {
	if a.dirtyManager == nil {
		return nil
	}
	items := a.dirtyManager.GetAllItems()
	var result []DirtyState
	for _, item := range items {
		result = append(result, DirtyState{
			ID:                  item.ID,
			IsDirty:             item.IsDirty,
			MinutesSinceLastSave: a.dirtyManager.GetMinutesSinceLastSave(item.ID),
		})
	}
	return result
}

func (a *App) ClearDirtyItem(id string) {
	if a.dirtyManager == nil {
		return
	}
	a.dirtyManager.ClearItem(id)
}

File: src-tauri/src/services/dirty_manager.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::sync::RwLock;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DirtyItem {
    pub id: String,
    #[serde(rename = "type")]
    pub item_type: String,
    pub original_hash: String,
    pub current_hash: String,
    pub is_dirty: bool,
    pub last_modified: DateTime<Utc>,
    pub last_saved: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DirtyState {
    pub id: String,
    pub is_dirty: bool,
    pub minutes_since_last_save: i64,
}

pub struct DirtyManager {
    items: RwLock<HashMap<String, DirtyItem>>,
}

impl DirtyManager {
    pub fn new() -> Self {
        Self {
            items: RwLock::new(HashMap::new()),
        }
    }
    fn compute_hash(content: &[u8]) -> String {
        let mut hasher = Sha256::new();
        hasher.update(content);
        format!("{:x}", hasher.finalize())
    }
    pub fn set_original(&self, id: &str, item_type: &str, content: &[u8]) {
        let hash = Self::compute_hash(content);
        let now = Utc::now();
        let mut items = self.items.write().unwrap();
        items.insert(
            id.to_string(),
            DirtyItem {
                id: id.to_string(),
                item_type: item_type.to_string(),
                original_hash: hash.clone(),
                current_hash: hash,
                is_dirty: false,
                last_modified: now,
                last_saved: now,
            },
        );
    }
    pub fn update_current(&self, id: &str, content: &[u8]) -> bool {
        let mut items = self.items.write().unwrap();
        if let Some(item) = items.get_mut(id) {
            let new_hash = Self::compute_hash(content);
            if new_hash != item.current_hash {
                item.current_hash = new_hash.clone();
                item.is_dirty = new_hash != item.original_hash;
                item.last_modified = Utc::now();
                return true;
            }
        }
        false
    }
    pub fn mark_saved(&self, id: &str, content: &[u8]) {
        let mut items = self.items.write().unwrap();
        if let Some(item) = items.get_mut(id) {
            let hash = Self::compute_hash(content);
            item.original_hash = hash.clone();
            item.current_hash = hash;
            item.is_dirty = false;
            item.last_saved = Utc::now();
        }
    }
    pub fn is_dirty(&self, id: &str) -> bool {
        let items = self.items.read().unwrap();
        items.get(id).map(|item| item.is_dirty).unwrap_or(false)
    }
    pub fn get_item(&self, id: &str) -> Option<DirtyItem> {
        let items = self.items.read().unwrap();
        items.get(id).cloned()
    }
    #[allow(dead_code)]
    pub fn get_dirty_items(&self) -> Vec<DirtyItem> {
        let items = self.items.read().unwrap();
        items.values().filter(|item| item.is_dirty).cloned().collect()
    }
    pub fn get_all_items(&self) -> Vec<DirtyItem> {
        let items = self.items.read().unwrap();
        items.values().cloned().collect()
    }
    pub fn clear_item(&self, id: &str) {
        let mut items = self.items.write().unwrap();
        items.remove(id);
    }
    pub fn has_any_dirty(&self) -> bool {
        let items = self.items.read().unwrap();
        items.values().any(|item| item.is_dirty)
    }
    pub fn get_minutes_since_last_save(&self, id: &str) -> i64 {
        let items = self.items.read().unwrap();
        items
            .get(id)
            .map(|item| Utc::now().signed_duration_since(item.last_saved).num_minutes())
            .unwrap_or(0)
    }
}

impl Default for DirtyManager {
    fn default() -> Self {
        Self::new()
    }
}

use std::sync::OnceLock;
static DIRTY_MANAGER: OnceLock<DirtyManager> = OnceLock::new();

fn get_manager() -> &'static DirtyManager {
    DIRTY_MANAGER.get_or_init(DirtyManager::new)
}

#[tauri::command]
pub fn get_dirty_state(id: String) -> Option<DirtyState> {
    let manager = get_manager();
    manager.get_item(&id).map(|item| DirtyState {
        id: item.id,
        is_dirty: item.is_dirty,
        minutes_since_last_save: manager.get_minutes_since_last_save(&id),
    })
}

#[tauri::command]
pub fn set_original_content(id: String, item_type: String, content: String) {
    get_manager().set_original(&id, &item_type, content.as_bytes());
}

#[tauri::command]
pub fn update_current_content(id: String, content: String) -> bool {
    get_manager().update_current(&id, content.as_bytes())
}

#[tauri::command]
pub fn mark_content_saved(id: String, content: String) {
    get_manager().mark_saved(&id, content.as_bytes());
}

#[tauri::command]
pub fn is_content_dirty(id: String) -> bool {
    get_manager().is_dirty(&id)
}

#[tauri::command]
pub fn has_any_dirty_content() -> bool {
    get_manager().has_any_dirty()
}

#[tauri::command]
pub fn get_all_dirty_states() -> Vec<DirtyState> {
    let manager = get_manager();
    manager
        .get_all_items()
        .into_iter()
        .map(|item| DirtyState {
            id: item.id.clone(),
            is_dirty: item.is_dirty,
            minutes_since_last_save: manager.get_minutes_since_last_save(&item.id),
        })
        .collect()
}

#[tauri::command]
pub fn clear_dirty_item(id: String) {
    get_manager().clear_item(&id);
}
